iree/compiler/Dialect/HAL/Transforms/CSEVariableLoads.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+class CSEVariableLoadsPass
+    : public PassWrapper<CSEVariableLoadsPass, OperationPass<FuncOp>> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+
+    Block *entryBlock = &funcOp.body().front();
+    auto entryBlockBuilder = OpBuilder::atBlockBegin(entryBlock);
+
+    DenseMap<StringRef, SmallVector<VariableLoadOp, 4>>
+        loadOpsGroupedByVariable;
+    funcOp.walk([&](VariableLoadOp loadOp) {
+      // Note: we assume that no IsolatedFromAbove regions are used with HAL
+      // variables here. If there are any, they would need to be treated
+      // independently of other regions.
+      loadOpsGroupedByVariable[loadOp.variable()].push_back(loadOp);
+    });
+    DenseSet<StringRef> storeOps;
+    funcOp.walk(
+        [&](VariableStoreOp storeOp) { storeOps.insert(storeOp.variable()); });
+    // TODO(scotttodd): check for VariableStoreIndirectOps?
```

Comment:
yes - also do that

Reply:
I'm not quite sure how to check if an indirect store can affect a specific variable. Could I just bail if _any_ indirect store is encountered?

For example, with this IR:
https://github.com/google/iree/blob/6f6336254ff73855839366d0c884447291d14b62/iree/compiler/Dialect/HAL/IR/test/variable_ops.mlir#L63-L73
I could follow `%1` through the `hal.variable.address` to get the variable name... as long as the `iree.ptr` is only directly used like that. Looks like it supports more general use that could be harder to reason about though?